ports/stm/common-hal/rgbmatrix/RGBMatrix.c

```diff
@@ -36,6 +36,7 @@ extern void _PM_IRQ_HANDLER(void);
 void *common_hal_rgbmatrix_timer_allocate() {
     TIM_TypeDef * timer = stm_peripherals_find_timer();
     stm_peripherals_timer_reserve(timer);
+    stm_peripherals_timer_never_reset(timer);
```

Comment:
with "never_reset", does `stm_peripherals_timer_free` in `common_hal_rgbmatrix_timer_free()` remain enough to fully deallocate/free/reset the timer?

Reply:
I'm sorry, can you rephrase? `stm_peripherals_timer_free` does undo never reset, is that your question? What do you mean by "remain"?

Reply:
![CopyQ NkqthX](https://user-images.githubusercontent.com/1517291/93711670-ba7f2080-fb15-11ea-8606-23a8fcf90e5e.png)

I'm still not fully familiar/comfortable with the transitions among these states; the naming is a bit different in each port and for each type of managed resource and I'm never entirely confident the semantics are the same.  (As a larger issue, it wouldn't hurt to document this for developers reference and to ensure consistency)

My question is whether the state transtion from "timer is in never_reset state" to "timer is in free state" by calling "timer_free" is a correct transition.  I think you answered yes.

Reply:
Yes. I think the going concept here is that `never_reset` is actually a wholly separate option from `reserved` - it is possible to have an un-`reserved` pin that is not in `never_reset`. I suggested joining the two at one point, so that `never_reset` was strictly a subset of `reserved`, but I believe we decided against it (I don't fully remember the reason).

Conceptually, `reserved` is the idea that a pin/peripheral cannot be claimed twice within a single runtime. `never_reset` is the idea that resetting will not change the `reserved` status, or the pin attributes of that pin/peripheral (basically, immunity status to the `reset_all` function). To turn off `never_reset`, you typically call the single-reset function (the not reset-all one) which completely resets all aspects of the pin/peripheral. We have some `reset_ok` functions which specifically turn off `never_reset` and nothing else, but they're rarely implemented and even more rarely used.

I'd definitely be down to document this or remove any inconsistencies that you have found. I think we could also revisit whether the two ideas should be combined more concretely in some way - basically, is there ever a scenario where you want something to be in never_reset without it actually being claimed?